### PR TITLE
feat: Add a plugins manifest to allow for more configuration of plugins

### PR DESCRIPTION
**Pull request by spacemaison**
* feat: Plugins can now be instantiated as headless (paneless)

* feat: Config field is passed through from manifest into plugins

### CODE REVIEW

**Comment by TheLostLambda** [comment]
Hey @spacemaison ! Sorry for letting this sit around for so long! This looks like a quite promising PR, but, as you point out in your issue on the topic, there are a lot of design decisions to be made here! For example, @imsnif @a-kenji and I have recently come to the decision that we want to somewhat unify the config and layout files (so layouts should be able to contain configuration like keybindings and also this sort of plugin configuration). How exactly this all happens, however, is a bit up in the air still.

I agree also that plugins really need a better way to have a default configuration. We've tossed around the idea of making plugins ZIP files with the WASM and a manifest within them (a bit like a Java `.jar`) or just using a `Default` implementation in the code. It seems a bit odd to me that `headless` plugins need to be configured as such when it really doesn't make sense for them to ever run in a pane, for example. We've previously thought about those implementing a different trait since `Plugin` contains `render` and that really doesn't make much sense.

Overall, there is a lot to talk about here! If possible, I'd love to have you join us for tomorrow's maintainer meeting to chat about it over a video call, but I'd also understand if that's not something you're comfortable with. If you can make the meeting (18:00-19:00 BST), then just ping either myself or @imsnif probably on the Discord!

Hopefully chat soon, otherwise we can do some writing to think about an ideal configuration API!

**Comment by spacemaison** [comment]
@TheLostLambda

Sure I'll try and make it tomorrow. I'll ping you in the morning before the meeting.

**Comment by TheLostLambda** [comment]
Hi @spacemaison ! It was nice chatting! As discussed offline, if you're up for removing the Default config stuff that might be best for the moment, and does this new system require every plugin to be listed in the config file before use in a layout?

If I understand correctly, this would break layouts like this?
https://github.com/zellij-org/rust-plugin-template/blob/main/plugin.yaml
(That layout is actually a bit outdated for a couple of other reasons, but I do know that the ability to specify paths in the layout is still something that people are relying on)

Looking forward to getting this merged soon!

**Comment by spacemaison** [comment]
It was nice chatting with you as well. From what I recall, I thought I added a way to load plugins directly from the layout files as well, but maybe I'm misremembering that here. I'm busy for the rest of the day but I'll try and look into it tonight and work on it then.

**Comment by spacemaison** [comment]
Hey @TheLostLambda, you're right I did just abstract away looking up plugins to using a tag name. Speaking with you all about it really helped clarify how handy it might be to have the ability to load and configure plugins directly from the layout files, and as it stands this PR breaks that workflow. @imsnif suggested that we use URL's for plugin lookup and that seems like a good way to support both. For the plugin template we can just use a file URL like so:

```yaml
---
direction: Horizontal
parts:
  - direction: Vertical
    split_size:
      Fixed: 1
    plugin: tab-bar
  - direction: Vertical
    plugin: "file:target/wasm32-wasi/debug/{{project-name}}.wasm"
  - direction: Vertical
    split_size:
      Fixed: 2
    plugin: status-bar
```

For plugin lookup from the manifest I suggest we just add a custom `zellij:` protocol for the URL's. So if you wanted to load the tab-bar plugin you would just specify `plugin: "zellij:tab-bar"`. For this PR we can just limit the valid protocols to those two, or we could expand the scope of this thing and load http and https URL's as well. What do you think?

**Comment by spacemaison** [comment]
I should also add that we don't have to use URL's here. I could just use the tag lookup and then fallback to trying to directly looking up the PathBuf as well. URL's just seemed like an elegant way to support standalone plugin lookup in layout files.

**Comment by TheLostLambda** [comment]
Hi @spacemaison ! I personally like the idea of starting to use URLs. With the changes you are proposing, using the default plugins could still just be something like `zellij:strider`? If so, go for it!

**Comment by a-kenji** [comment]
Just to add, since I don't know how involved it would be. 
I think it should be possible to add the configuration also directly in the layout itself.
```
---
direction: Horizontal
parts:
  - direction: Vertical
    split_size:
      Fixed: 1
    run:
      plugin: 
        path: "zellij:tab-bar"
        config:  ""
  - direction: Vertical
     run:
        plugin: 
          path: "file:target/wasm32-wasi/debug/{{project-name}}.wasm"
  - direction: Vertical
    split_size:
      Fixed: 2
    run:
      plugin: 
        path: "status-bar"
```

**Comment by spacemaison** [comment]
Absolutely @a-kenji, I was going to back out of adding the config field to this PR and then start a separate issue for it so we can be sure to get it right.

@TheLostLambda I've refactored everything to use URL's so the plugin templates will work again. I've just about finished rebasing the recent permissions changes into this PR as well, but it's super late for me now so I'm going to finish it off in the morning.

**Comment by a-kenji** [comment]
@spacemaison 
No problem, just wanted to make sure it's still in the back of our heads!
Putting it in a separate pr sounds good.

**Comment by TheLostLambda** [comment]
@spacemaison Thanks for keeping at this! Once you feel it's all ready for review, feel free to ping me on Discord!

**Comment by spacemaison** [comment]
Hey @TheLostLambda, apparently I am the midnight coder here. I've finished rebasing main on top of this PR. I'll try and ping you in the morning on Discord to let you know as well.

File: zellij-utils/src/input/plugins.rs
Comment (TheLostLambda) [inline zellij-utils/src/input/plugins.rs:148]: Agreed! I think this might be helpful as another type of `Pane` as opposed to just a plugin type :) Happy to chat about it down the line!

**Review by TheLostLambda — APPROVED** [review]
Overall looks great! I left a couple of nitpicks if you are interested, but I'm happy to merge this whenever you'd like me to!

**Comment by TheLostLambda** [comment]
@spacemaison Oh, actually, could we fix this? 
image


The pane frames get their names from the plugin URL, it might be worth stripping away the `zellij:` and just using the tag name. For `file:` it's probably fine (and good) to display it's a non-default plugin (though I suppose you could also strip the `file:` and just display the path there.

**Comment by TheLostLambda** [comment]
Whoops, okay, maybe I didn't test enough :sweat_smile: Running commands also seems broken? Even when `_allow_exec_host_cmd` is set, it seems to deny permissions. I think that `run._allow...` is correctly set, but that must get overridden somehow when turning it into a `Plugin`
image

This is the test layout I was using:
```yaml
---
template:
  direction: Horizontal
  parts:
    - direction: Vertical
      borderless: true
      split_size:
        Fixed: 1
      run:
        plugin:
          location: "zellij:tab-bar"
    - direction: Vertical
      body: true
    - direction: Vertical
      borderless: true
      split_size:
        Fixed: 2
      run:
        plugin:
          location: "zellij:status-bar"
tabs:
  - direction: Vertical
    parts:
      - direction: Horizontal
        split_size:
          Percent: 20
        run:
          plugin:
            location: "zellij:strider"
            _allow_exec_host_cmd: true
      - direction: Horizontal
```

This lets you see the permission messages: `tail -f /tmp/zellij-1000/zellij-log/zellij.log`

File: zellij-server/src/wasm_vm.rs
Comment (spacemaison) [reply zellij-server/src/wasm_vm.rs:57]: Sure, you're right it's confusing.

File: zellij-server/src/wasm_vm.rs
Comment (spacemaison) [reply zellij-server/src/wasm_vm.rs:89]: Yeah the "when" and "where" of starting headless plugins was something I took into consideration. Starting them right after the wasm thread is forked was just the simplest and most obvious choice for a first pass at it. I absolutely could start them via `PluginInstruction::Load` message, but I'm worried it might just obfuscate what's happening. Right after the call to `init_session` in `zellij-server/src/lib.rs` maybe? Is there a reason you want it changed aside from the aesthetics of it being a bit ugly?

In any event I moved the security warning into the call to `start_plugin` so that headless plugins get the warning as well.

File: zellij-server/src/wasm_vm.rs
Comment (spacemaison) [reply zellij-server/src/wasm_vm.rs:233]: Sure thing

File: zellij-server/src/wasm_vm.rs
Comment (spacemaison) [reply zellij-server/src/wasm_vm.rs:246]: Thanks for catching this, it was a hold over from the first pass at this when I still called them service plugins

File: zellij-utils/src/input/layout.rs
Comment (spacemaison) [reply zellij-utils/src/input/layout.rs:480]: Yup, it is. I think I was thinking of `Result::flatten` or something. I changed it, thanks

File: zellij-utils/src/input/plugins.rs
Comment (spacemaison) [reply zellij-utils/src/input/plugins.rs:48]: Yeah more or less. That's where the run configuration from the layout files meets the plugins configuration, the FIXME tag was more of a general reminder to not screw up merging configuration metadata from the layout files, which i promptly ignored and did anyways. :P

File: zellij-utils/src/input/plugins.rs
Comment (spacemaison) [reply zellij-utils/src/input/plugins.rs:148]: Yup! Someone in the Discord was doing a pomodoro timer in the Discord the other day, and they we're asking for what was essentially a "static" plugin pane, or a plugin that runs once and mirrors itself to every pane. I'll open an issue to discuss it after this gets merged.

**Comment by spacemaison** [comment]
I fixed displaying the full URL in the pane frames by just displaying the paths or the URL's.

The permission for executing commands should be fixed as well. I was copying over the permission from the layout config when generating the plugin config for plugins with the `zellij:` URL protocol. I added a test to assert that it's doing the correct thing now.

**Comment by TheLostLambda** [comment]
@spacemaison I'm happy to merge things when you are ready! It actually looks like resolving those conflicts was slightly more complex than I thought (something I probably shouldn't try in the Github web editor :sweat_smile:), so it would be great if you could take a swing at them! Otherwise I can do so tonight or tomorrow :)

**Comment by spacemaison** [comment]
Hey, sure I'll get to it over the next few hours.

**Comment by TheLostLambda** [comment]
@spacemaison Thanks for the fixes! I actually wanted to change those data directories to be the full paths, but I must have missed that in my earlier reviewing! Thanks for tweaking that!

I'll add some info to the changelog :)

Thank you for all of your hard work here!

**Pull request merged**
